### DISCUSSION TITLE
Remapping keys with Karabiner Elements and Hammerspoon

### DISCUSSION THREAD

**Discussion by jannis-baum** [godbout/kindaVim.blahblah#117] [Show and tell]
The DistributedNotification API introduced in [kV 33](https://github.com/godbout/kindaVim.docs/releases/tag/33) allows us to remap keys by modes with the help of [Karabiner Elements](https://karabiner-elements.pqrs.org) and [Hammerspoon](http://www.hammerspoon.org).

I will use this thread to illustrate this at the example of **remapping `q` in Normal mode in Safari to open a new tab**.

### Setting up a Karabiner variable to reflect kV's mode

The first step is getting Karabiner to know whether kV is in Normal mode. This is done by listening to the `kindaVimDidEnterNormalMode` and `kindaVimDidExitNormalMode` notifications with Hammerspoon and using Karabiner's CLI to set a Karabiner variable accordingly. Karabiner's CLI can be found in `/Library/Application\\ Support/org.pqrs/Karabiner-Elements/bin/karabiner_cli` but may be at a different location depending on your installation. Add the following to your Hammerspoon config, e.g. in `~/.hammerspoon/init.lua`:

```lua
kVNormalEnterWatcher = hs.distributednotifications.new(function(name, object, userInfo)
    hs.execute('/Library/Application\\ Support/org.pqrs/Karabiner-Elements/bin/karabiner_cli --set-variables "{\\"kVNormal\\":1}"')
end, 'kindaVimDidEnterNormalMode')
kVNormalEnterWatcher:start()

kVNormalExitWatcher = hs.distributednotifications.new(function(name, object, userInfo)
    hs.execute('/Library/Application\\ Support/org.pqrs/Karabiner-Elements/bin/karabiner_cli --set-variables "{\\"kVNormal\\":0}"')
end, 'kindaVimDidExitNormalMode')
kVNormalExitWatcher:start()
```

With this config loaded, Karabiner will have the variable `kVNormal` set to `1` or `0` depending on whether kV is in normal mode. You can observe the variable with Karabiner's EventViewer App to verify everything is working correctly.

### Remapping the key

With this variable working, you can use it as a condition in a [complex modification](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/) in Karabiner. The easiest way to configure Karabiner is using [goku](https://github.com/yqrashawn/GokuRakuJoudo), which provides more concise `.edn` syntax over Karabiner's `.json`. The following snippet in your `~/.config/karabiner.edn` is enough to achieve our goal of **remapping `q` in Normal mode in Safari to open a new tab**, i.e. to `cmd+t`.

```edn
{
  :applications {
    :safari ["^com\\.apple\\.Safari$"]}

  :main [
    {:des "safari normal mode" :rules [
      [:condis :safari ["kVNormal" 1]]
        [:q :!Ct]]}
  ]
}
```

With this in your `karabiner.edn`, run `goku` and your remapping will be live!

**Note**: Make sure you have a Karabiner profile named `Default` or set up a different profile name with goku.

**Note 2**: For `command+t` to actually arrive at Safari, you need to enable *Passthrough* for `command` in kV's General preferences.

**Comment by haveanicedavid** [comment]
whoa - this is awesome! You could make entire app-specific utilities - i’ll definitely be looking into this more later

**Reply by jannis-baum** [reply]
> i’ll definitely be looking into this more later

Share if you come up with something cool!
